### ISSUE TITLE
Cannot eager load multiple eagers off of the same relation

### ISSUE THREAD

**Issue by josephmancuso** [MasoniteFramework/orm#233]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%240%20Funded-%2300A156.svg" />](https://issuehunt.io/r/MasoniteFramework/orm/issues/233)

Need to add eager loading support. This feature must have support for nested relationships to eager loading.

Currently there seems to be an issue with eager loading a relationship off the same eager loading relationship.

For example we can't do this:

```python
User.with_('articles.logo', 'articles.comments').get()
```

It doesn't seem to fetch properly. Doing a single eager load like `articles.comments.author` works fine but when we use eager load another nested relation from articles such as `articles.author`, it overwrite the original relationship.

The code is located in the QueryBuilder classes `prepare_result` method. This is where the eager loading is called and registered to the model

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/MasoniteFramework/orm/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by Marlysson** [comment]
@josephmancuso In future we can change the syntax to receive a list instead of append the model name in every relation?

```python
User.with("comments", {"articles":["logo", "comments"]})
```

**Comment by josephmancuso** [comment]
We need the relationship name. The example is eager loading articles, the articles logos as well as the articles comments

**Comment by josephmancuso** [comment]
The example you gave would just be eager loading the users logo and users comments

**Comment by Marlysson** [comment]
Yeah, I've updated the example, but turns into something very complex.

**Comment by josephmancuso** [comment]
Yeah I don't think that improves anything. Makes it complicated.

**Comment by josephmancuso** [comment]
I'd be open to doing something like `with_('articles.logo,comments')`

**Comment by josephmancuso** [comment]
But let's get this working first lol

**Comment by josephmancuso** [comment]
@Marlysson I actually think you are right. The data structure we have is not going to work. The issue is that if we do something like:

```python
User.with('articles.comments', 'articles.author') 
```

It will register the `articles.comments` relationship on the user but then it gets overridden when we add the a `articles.author` relationship because its both registered to the user model.

@Marlysson So I think we need a different data structure here. I want to keep `articles.comment, articles.author` but I think inside we need to compile it down to:

```python
{"articles": ['comments', 'author']}
```

This would also make single relationships something like this:

```python
{"comments": ["comments"]}
```

but we can deal with that.

**Comment by josephmancuso** [comment]
Having a hard time with this 😞 . Need to keep trying different things here

**Comment by RoySegall** [comment]
I know that the Drupal 7 Entity module have something which calls `EntityMetadtaWrapper` which does something like:
```php
$user->favorate_food->get(0)->label() // Will get the title which the user mark as favorate.
```
There's can be something more complex like:
```php
$topic->author->groups->get(0)->label() // Will get the first groups user belongs to
```

This is done by using getters and setters on the object and loads. Is that the same case with eager loading?

**Comment by josephmancuso** [comment]
@RoySegall that's not exactly eager loading I believe. That is more relationships and relationship chaining (which the ORM supports already).

How eager loading works here is inside the queryBuilder before the result is returned, If there are any relationships registered for eager loading (using `with_('articles')`) then it needs to load the full relationship into memory and register it to the user model. 

It works exactly as it should now but the issue is that when you eager load the same relation again it overwrites the existing one.

for example if you eager load `User.with_('articles.user', 'articles.comments')` it will fetch the relation the articles relation as well as the user relation on the articles object perfectly fine and registers it to the user model. But when it goes to eager load the second one (articles.comments) it registers it to the user model and overwrites that original articles registration to the user model. So now the articles model that has loaded the user object is now gone.

The solution I believe is to structure the for loop in such a way where it loads everything it needs to on the articles relation AND THEN registers it to the user model. So it prevents requiring 2 registrations to the user model. This will also prevent 2 fetches between the user model and articles model.

It might be a bit confusing but the concept of eager loading is pretty high level

**Comment by Marlysson** [comment]
@josephmancuso The current trying is in feature/233 branch?

**Comment by josephmancuso** [comment]
@Marlysson yes but I've ripped out and rewrote it a bunch of different ways so there's been no real progress on it

**Comment by RoySegall** [comment]
any change for a DB for example or on how to set the records to get the bug? Would be more useful when trying to reproduce and fix the bug

**Comment by josephmancuso** [comment]
Feel free to modify the orm.sqlite3 database in the root of the project. Can use a normal database client like table plus.

To replicate the issue at the database level would be to eager load the same root relation twice.

So we have a relationship currently that goes:

```
User --<- Articles >-- Logo
User --<- Articles
```

In other words, 

* User has many articles and an articles has one logo
* An article has one user

This relationship is setup in that sqlite database so you can hop in that sqlite test_eager_loading test.

So when we try to do this in the code. That logo and user should both be registered to the article model. then that article model registers to the User model. So now that user model has the articles relationship. When we access the article relationship on the user model, it pulls this registered relationship instead of calling the database (because we already fetched and registered the relationship so no need to go to the database).

With that being said. a test like this fails:

```
user = User.with_('articles.logo', 'articles.user').get().serialize()
self.assertTrue(user.articles[0]['logo'])
self.assertTrue(user.articles[0]['user'])
```

It fails because in the prepare_result for loop, it is overriding the articles.user relationship (which works perfectly fine) with the article.logo one we registered before.

In other words. This works fine:

```
# test 1
user = User.with_('articles.logo').get().serialize()
self.assertTrue(user.articles[0]['logo'])
```

```
# test 2
user = User.with_('articles.user').get().serialize()
self.assertTrue(user.articles[0]['user'])
```

But it does not work together because 1 overwrites the other

**Comment by josephmancuso** [comment]
I completed this myself and cancelled funding for it. Thank you for everyone interested in this issue! ❤️
